File: D_PriceFixed.cpp
#include <cstdio>
#include <algorithm>
#include <vector>
#include <utility>
using namespace std;
int main(){
    int n;
    long long a, b, needed=0;

    scanf("%d", &n);
    vector< pair<long long, long long> > v;
    for(int i = 0;i < n;i++){
        scanf("%lld %lld", &a, &b);
        needed += a;
        v.push_back(make_pair(b, a));
    }
    sort(v.begin(), v.end());

    long long items=0, total=0, buy;
    int i=0, j=n-1;

    while(items < needed){
        if(v[i].first <= items){
            buy = v[i].second;
            v[i].second = 0;
            items += buy;
            total += buy;
            i++;
        } else {
            if(v[j].second >= v[i].first - items){
                buy = v[i].first - items;
                v[j].second -= buy;
                items += buy;
                total += buy*2;
                if(v[j].second == 0){
                    j--;
                }
            } else {
                buy = v[j].second;
                v[j].second = 0;
                items += buy;
                total += buy*2;
                j--;
            }
        }
    }
    printf("%lld", total);
}

File: D_PriceFixed.py
n = int(input())
l = []
needed = 0
for _ in range(n):
    a, b = map(int, input().split())
    needed += a
    l.append(([b, a]))
l.sort()
items, total = 0, 0
i, j = 0, n-1
while(items < needed):
    if l[i][0] <= items:
        buy = l[i][1]
        l[i][1] = 0
        items += buy
        total += buy
        i += 1
    else:
        if l[j][1] >= l[i][0] - items:
            buy = l[i][0] - items
            l[j][1] -= buy
            items += buy
            total += buy*2
            if l[j][1] == 0:
                j -= 1
        else:
            buy = l[j][1]
            l[j][1] = 0
            items += buy
            total += buy*2
            j -= 1
print(total)